src/plugins/borrow-plugins/types.js

```diff
@@ -0,0 +1,111 @@
+// @flow
+import { type EdgeAccount, type EdgeCurrencyWallet, type EdgeNetworkFee } from 'edge-core-js'
+
+// Proposal for a "service plugin" which is a Higher-Order-Plugin that can be
+// instantiated with some account context and returns a service instance.
+export type EdgeServicePlugin<Info, Api> = {
+  info: Info,
+  init: (account: EdgeAccount) => Promise<Api>
+}
+
+// Borrow more:
+export type BorrowRequest = {
+  tokenId?: string,
+  nativeAmount: string,
+
+  // Optional wallet which will deposit on behalf of the loan account mainWallet
+  fromWallet?: string
+}
+
+// Make payment:
+export type RepayRequest = {
+  tokenId?: string,
+  nativeAmount: string,
+
+  // Optional wallet which will deposit on behalf of the loan account mainWallet
+  fromWallet?: string
+}
+
+// Deposit collateral:
+export type DepositRequest = {
+  tokenId?: string,
+  nativeAmount: string,
+
+  // Optional wallet which will deposit on behalf of the loan account mainWallet
+  fromWallet?: string
+}
+
+// Withdraw collateral:
+export type WithdrawRequest = {
+  tokenId?: string,
+  nativeAmount: string,
+
+  // Optional destination for the funds
+  toWallet?: string
+}
+
+// General purpose approvable action
+export type ApprovableAction = {
+  +networkFee: EdgeNetworkFee,
+  +approve: () => Promise<void>
+}
+
+//
+// Loan Account – Primary entity
+//
+
+export type LoanAccount = {
+  // Account wallet for the loan (e.g. Ethereum wallet)
+  mainWallet: EdgeCurrencyWallet,
+
+  collaterals: Array<{
+    tokenId?: string,
+    nativeAmount: string
+  }>,
+  debts: Array<{
+    tokenId?: string,
+    nativeAmount: string,
+    apy: number
+  }>,
+
+  loanToValue: number,
+
+  getApr: (tokenId?: string) => Promise<number>,
+
+  // Collateral modification
+  deposit: (request: DepositRequest) => Promise<ApprovableAction>,
+  withdraw: (request: WithdrawRequest) => Promise<ApprovableAction>,
+
+  // Debt modification
+  borrow: (request: BorrowRequest) => Promise<ApprovableAction>,
+  repay: (request: RepayRequest) => Promise<ApprovableAction>,
+
+  // Close entire account
+  close: () => Promise<ApprovableAction>
+}
+
+//
+// Plugin Info
+//
+
+export type BorrowPluginInfo = {
+  pluginId: string,
+
+  // Maximum loan-to-value ratio
+  maxLtvRatio: number
+}
+
+//
+// Service
+//
+
+export type BorrowService = {
+  fetchLoanAccount: (wallet: EdgeCurrencyWallet) => Promise<LoanAccount>,
+  fetchLoanAccounts: () => Promise<LoanAccount[]>
```

Comment:
We need to think about how loan plugins relate to currency plugins. I do not think the plugins should be responsible for this.

One idea would be to simply create a loan manager for every wallet in an account, for each supported plugin type. If Edge supports 5 loan contracts on ETH, then each ETH wallet would get 5 loan managers. Most would have a 0 balance, which is fine. If the user uses an external app + wallet connect to start a new loan, then Edge we will see that on next login, because we always check all possible contracts.

Another idea is to keep track of which walletId's have which loan types on them. This won't auto-detect external loans, but could save resources.

Yet another idea is to only open the plugin when the user opens the loan scene. This would combine the reliability of the first idea with the efficiency of the second idea.

In any case, it should *not* be the responsibility of this plugin to know which wallets it belongs to. The core / GUI should be in charge of that, since there are different possible strategies, and because it's common logic we shouldn't have to replicate between plugins.

Reply:
Ah I see. We should discuss these options together.